include/graphblas/algorithms/triangle_count.hpp

```diff
@@ -0,0 +1,249 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * @file
+ *
+ * Implements the triangle counting algorithm, using different methods.
+ *
+ * @author Benjamin Lozes
+ * @date: May 10th, 2023
+ */
+
+#ifndef _H_GRB_TRIANGLE_COUNT
+#define _H_GRB_TRIANGLE_COUNT
+
+#include <map>
+#include <numeric>
+#include <vector>
+
+#include <graphblas/utils/iterators/nonzeroIterator.hpp>
+
+#include <graphblas.hpp>
+
+namespace grb {
+
+	namespace algorithms {
+
+		enum class TriangleCountAlgorithm { Burkhardt, Cohen, Sandia_TT };
+
+		std::map< TriangleCountAlgorithm, std::string > TriangleCountAlgorithmNames = {
+			{ TriangleCountAlgorithm::Burkhardt, "Burkhardt" },
+			{ TriangleCountAlgorithm::Cohen, "Cohen" },
+			{ TriangleCountAlgorithm::Sandia_TT, "Sandia_TT" }
+		};
+
+		template<
+			class Semiring, class MulMonoid, class SumMonoid,
+			Descriptor descr_mxm = descriptors::no_operation,
+			Descriptor descr_ewa = descriptors::no_operation,
+			Descriptor descr_reduce = descriptors::no_operation,
+			typename D1, typename RIT1, typename CIT1, typename NIT1,
+			typename D2, typename RIT2, typename CIT2, typename NIT2,
+			typename D3, typename RIT3, typename CIT3, typename NIT3,
+			typename D4, typename RIT4, typename CIT4, typename NIT4,
+			typename D5, typename RIT5, typename CIT5, typename NIT5,
+			typename D6
+		>
+		RC triangle_count_generic(
+			size_t & count,
+			Matrix< D1, config::default_backend, RIT1, CIT1, NIT1 > & MXM_out,
+			const Matrix< D2, config::default_backend, RIT2, CIT2, NIT2 > & MXM_lhs,
+			const Matrix< D3, config::default_backend, RIT3, CIT3, NIT3 > & MXM_rhs,
+			Matrix< D4, config::default_backend, RIT4, CIT4, NIT4 > & EWA_out,
+			const Matrix< D5, config::default_backend, RIT5, CIT5, NIT5 > & EWA_rhs,
+			const D6 div_factor,
+			const Semiring mxm_semiring = Semiring(),
+			const MulMonoid ewiseapply_monoid = MulMonoid(),
+			const SumMonoid sumreduce_monoid = SumMonoid()
+		) {
+			if( ( &MXM_out == &MXM_lhs ) || ( &MXM_out == &MXM_rhs ) ) {
+				return ILLEGAL;
+			}
+
+			RC rc = SUCCESS;
+
+			// Compute MXM_out = Mlhs * Mrhs
+			rc = rc ? rc : mxm< descr_mxm >( MXM_out, MXM_lhs, MXM_rhs, mxm_semiring, Phase::RESIZE );
+			rc = rc ? rc : mxm< descr_mxm >( MXM_out, MXM_lhs, MXM_rhs, mxm_semiring, Phase::EXECUTE );
+
+			// Compute MXM_out .*= EWA_rhs
+			// FIXME: Replace by a foldl( Matrix[in,out], Matrix[in], Monoid ) - not implemented yet
+			// Will then become:
+			// rc = rc ? rc : eWiseApply< descr_ewa >( MXM_out, MXM_out, EWA_rhs, ewiseapply_monoid, Phase::RESIZE );
+			// rc = rc ? rc : eWiseApply< descr_ewa >( MXM_out, MXM_out, EWA_rhs, ewiseapply_monoid, Phase::EXECUTE );
```

Comment:
eWiseApply -> foldl and remove one of the duplicate `MXM_out` arguments?

Reply:
Actually, if replacing it with `foldl`, then a `clear` on `MXM_out` becomes mandatory right?